File: app/src/main/java/com/talapp/mapme/util/ExportUtils.kt
package com.talapp.mapme.util

import android.content.Context
import android.content.Intent
import androidx.core.content.FileProvider
import com.talapp.mapme.data.Walk
import com.talapp.mapme.data.WalkPoint
import com.talapp.mapme.data.WalkPoi
import java.io.File
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone

/**
 * Utilities for serializing and exporting recorded walks to standard GPX 1.1 and KML 2.2 formats.
 */
object ExportUtils {

    /**
     * Generates a valid GPX 1.1 XML string from a recorded walk, path points, and waypoints.
     */
    fun generateGpxXml(walk: Walk, points: List<WalkPoint>, pois: List<WalkPoi>): String {
        val isoFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.US).apply {
            timeZone = TimeZone.getTimeZone("UTC")
        }
        val safeWalkTitle = walk.title.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;")

        val sb = StringBuilder()
        sb.append("<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n")
        sb.append("<gpx version=\"1.1\" creator=\"MapMe\" xmlns=\"http://www.topografix.com/GPX/1/1\">\n")
        sb.append("  <metadata>\n")
        sb.append("    <name>$safeWalkTitle</name>\n")
        sb.append("    <time>${isoFormat.format(Date(walk.startTime))}</time>\n")
        sb.append("  </metadata>\n")

        for (poi in pois) {
            val name = (poi.text ?: "Waypoint").replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;")
            sb.append("  <wpt lat=\"${poi.latitude}\" lon=\"${poi.longitude}\">\n")
            sb.append("    <name>$name</name>\n")
            sb.append("    <time>${isoFormat.format(Date(poi.timestamp))}</time>\n")
            sb.append("  </wpt>\n")
        }

        sb.append("  <trk>\n")
        sb.append("    <name>$safeWalkTitle</name>\n")
        sb.append("    <trkseg>\n")
        for (pt in points) {
            sb.append("      <trkpt lat=\"${pt.latitude}\" lon=\"${pt.longitude}\">\n")
            sb.append("        <time>${isoFormat.format(Date(pt.timestamp))}</time>\n")
            sb.append("        <speed>${pt.speed}</speed>\n")
            sb.append("      </trkpt>\n")
        }
        sb.append("    </trkseg>\n")
        sb.append("  </trk>\n")
        sb.append("</gpx>\n")

        return sb.toString()
    }

    /**
     * Generates a valid KML 2.2 XML string from a recorded walk, path points, and waypoints.
     */
    fun generateKmlXml(walk: Walk, points: List<WalkPoint>, pois: List<WalkPoi>): String {
        val safeWalkTitle = walk.title.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;")
        val sb = StringBuilder()
        sb.append("<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n")
        sb.append("<kml xmlns=\"http://www.opengis.net/kml/2.2\">\n")
        sb.append("  <Document>\n")
        sb.append("    <name>$safeWalkTitle</name>\n")
        sb.append("    <Style id=\"routeLine\">\n")
        sb.append("      <LineStyle>\n")
        sb.append("        <color>ff00ffff</color>\n") // AABBGGRR: cyan
        sb.append("        <width>5</width>\n")
        sb.append("      </LineStyle>\n")
        sb.append("    </Style>\n")

        // Waypoints (Placemarks)
        for ((idx, poi) in pois.withIndex()) {
            val name = (poi.text ?: "POI #${idx + 1}").replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;")
            sb.append("    <Placemark>\n")
            sb.append("      <name>$name</name>\n")
            sb.append("      <Point>\n")
            sb.append("        <coordinates>${poi.longitude},${poi.latitude},0</coordinates>\n")
            sb.append("      </Point>\n")
            sb.append("    </Placemark>\n")
        }

        // Track Polyline
        if (points.isNotEmpty()) {
            sb.append("    <Placemark>\n")
            sb.append("      <name>Track: $safeWalkTitle</name>\n")
            sb.append("      <styleUrl>#routeLine</styleUrl>\n")
            sb.append("      <LineString>\n")
            sb.append("        <tessellate>1</tessellate>\n")
            sb.append("        <coordinates>\n")
            val coordStr = points.joinToString(" ") { "${it.longitude},${it.latitude},0" }
            sb.append("          $coordStr\n")
            sb.append("        </coordinates>\n")
            sb.append("      </LineString>\n")
            sb.append("    </Placemark>\n")
        }

        sb.append("  </Document>\n")
        sb.append("</kml>\n")

        return sb.toString()
    }

    /**
     * Writes GPX XML to temporary cache file and launches Android share sheet.
     */
    fun exportGpx(context: Context, walk: Walk, points: List<WalkPoint>, pois: List<WalkPoi>) {
        try {
            val xml = generateGpxXml(walk, points, pois)
            val safeTitle = walk.title.replace("[^a-zA-Z0-9_-]".toRegex(), "_")
            val file = File(context.cacheDir, "${safeTitle}_${walk.id}.gpx")
            file.writeText(xml)

            val uri = FileProvider.getUriForFile(context, "${context.packageName}.fileprovider", file)
            val intent = Intent(Intent.ACTION_SEND).apply {
                type = "application/gpx+xml"
                putExtra(Intent.EXTRA_STREAM, uri)
                putExtra(Intent.EXTRA_SUBJECT, "MapMe Route: ${walk.title}")
                addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
            }
            context.startActivity(Intent.createChooser(intent, "Export GPX Route"))
        } catch (e: Exception) {
            android.widget.Toast.makeText(context, "Export error: ${e.message}", android.widget.Toast.LENGTH_SHORT).show()
        }
    }

    /**
     * Writes KML XML to temporary cache file and launches Android share sheet.
     */
    fun exportKml(context: Context, walk: Walk, points: List<WalkPoint>, pois: List<WalkPoi>) {
        try {
            val xml = generateKmlXml(walk, points, pois)
            val safeTitle = walk.title.replace("[^a-zA-Z0-9_-]".toRegex(), "_")
            val file = File(context.cacheDir, "${safeTitle}_${walk.id}.kml")
            file.writeText(xml)

            val uri = FileProvider.getUriForFile(context, "${context.packageName}.fileprovider", file)
            val intent = Intent(Intent.ACTION_SEND).apply {
                type = "application/vnd.google-earth.kml+xml"
                putExtra(Intent.EXTRA_STREAM, uri)
                putExtra(Intent.EXTRA_SUBJECT, "MapMe Route: ${walk.title}")
                addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
            }
            context.startActivity(Intent.createChooser(intent, "Export KML Route"))
        } catch (e: Exception) {
            android.widget.Toast.makeText(context, "Export error: ${e.message}", android.widget.Toast.LENGTH_SHORT).show()
        }
    }
}

// Top-level delegating functions for convenient imports
fun exportGpx(context: Context, walk: Walk, points: List<WalkPoint>, pois: List<WalkPoi>) {
    ExportUtils.exportGpx(context, walk, points, pois)
}

fun exportKml(context: Context, walk: Walk, points: List<WalkPoint>, pois: List<WalkPoi>) {
    ExportUtils.exportKml(context, walk, points, pois)
}

File: web/js/exportUtils.js
// ==========================================================================
// MapMe Web Dashboard - GPX & KML Exporters
// ==========================================================================

function escapeXml(unsafe) {
    if (!unsafe) return "";
    return unsafe.replace(/[<>&'"]/g, (c) => {
        switch (c) {
            case "<": return "&lt;";
            case ">": return "&gt;";
            case "&": return "&amp;";
            case "'": return "&apos;";
            case "\"": return "&quot;";
        }
    });
}

/**
 * Generates a valid GPX 1.1 XML string from a walk record, points, and POIs.
 */
export function generateGpxXml(walk, points, pois = []) {
    const startTimeIso = new Date(walk.startTime || Date.now()).toISOString();
    const safeTitle = escapeXml(walk.title || "MapMe Route");

    let xml = `<?xml version="1.0" encoding="UTF-8"?>\n`;
    xml += `<gpx version="1.1" creator="MapMe" xmlns="http://www.topografix.com/GPX/1/1">\n`;
    xml += `  <metadata>\n`;
    xml += `    <name>${safeTitle}</name>\n`;
    xml += `    <time>${startTimeIso}</time>\n`;
    xml += `  </metadata>\n`;

    // Waypoints
    for (const poi of pois) {
        const name = escapeXml(poi.text || "Waypoint");
        const poiTimeIso = new Date(poi.timestamp || walk.startTime).toISOString();
        xml += `  <wpt lat="${poi.latitude}" lon="${poi.longitude}">\n`;
        xml += `    <name>${name}</name>\n`;
        xml += `    <time>${poiTimeIso}</time>\n`;
        xml += `  </wpt>\n`;
    }

    // Track
    xml += `  <trk>\n`;
    xml += `    <name>${safeTitle}</name>\n`;
    xml += `    <trkseg>\n`;
    for (const pt of points) {
        const ptTimeIso = new Date(pt.timestamp || walk.startTime).toISOString();
        const speed = pt.speed != null ? pt.speed : 0;
        xml += `      <trkpt lat="${pt.latitude}" lon="${pt.longitude}">\n`;
        xml += `        <time>${ptTimeIso}</time>\n`;
        xml += `        <speed>${speed}</speed>\n`;
        xml += `      </trkpt>\n`;
    }
    xml += `    </trkseg>\n`;
    xml += `  </trk>\n`;
    xml += `</gpx>\n`;

    return xml;
}

/**
 * Generates a valid KML 2.2 XML string from a walk record, points, and POIs.
 */
export function generateKmlXml(walk, points, pois = []) {
    const safeTitle = escapeXml(walk.title || "MapMe Route");

    let xml = `<?xml version="1.0" encoding="UTF-8"?>\n`;
    xml += `<kml xmlns="http://www.opengis.net/kml/2.2">\n`;
    xml += `  <Document>\n`;
    xml += `    <name>${safeTitle}</name>\n`;
    xml += `    <Style id="routeLine">\n`;
    xml += `      <LineStyle>\n`;
    xml += `        <color>ff00ffff</color>\n`; // AABBGGRR: cyan
    xml += `        <width>5</width>\n`;
    xml += `      </LineStyle>\n`;
    xml += `    </Style>\n`;

    // Waypoints (Placemarks)
    for (let i = 0; i < pois.length; i++) {
        const poi = pois[i];
        const name = escapeXml(poi.text || `POI #${i + 1}`);
        xml += `    <Placemark>\n`;
        xml += `      <name>${name}</name>\n`;
        xml += `      <Point>\n`;
        xml += `        <coordinates>${poi.longitude},${poi.latitude},0</coordinates>\n`;
        xml += `      </Point>\n`;
        xml += `    </Placemark>\n`;
    }

    // Track Polyline
    if (points.length > 0) {
        xml += `    <Placemark>\n`;
        xml += `      <name>Track: ${safeTitle}</name>\n`;
        xml += `      <styleUrl>#routeLine</styleUrl>\n`;
        xml += `      <LineString>\n`;
        xml += `        <tessellate>1</tessellate>\n`;
        xml += `        <coordinates>\n`;
        const coordStr = points.map(pt => `${pt.longitude},${pt.latitude},0`).join(" ");
        xml += `          ${coordStr}\n`;
        xml += `        </coordinates>\n`;
        xml += `      </LineString>\n`;
        xml += `    </Placemark>\n`;
    }

    xml += `  </Document>\n`;
    xml += `</kml>\n`;

    return xml;
}

/**
 * Triggers a browser file download of string content.
 */
export function downloadFile(filename, content, mimeType) {
    const blob = new Blob([content], { type: mimeType });
    const url = URL.createObjectURL(blob);
    const a = document.createElement("a");
    a.href = url;
    a.download = filename;
    document.body.appendChild(a);
    a.click();
    document.body.removeChild(a);
    URL.revokeObjectURL(url);
}

/**
 * Convenience helper to export GPX file from walk data.
 */
export function exportGpx(walk, points, pois = []) {
    const xml = generateGpxXml(walk, points, pois);
    const safeName = (walk.title || "route").replace(/[^a-zA-Z0-9_-]/g, "_");
    downloadFile(`${safeName}_${walk.id || Date.now()}.gpx`, xml, "application/gpx+xml");
}

/**
 * Convenience helper to export KML file from walk data.
 */
export function exportKml(walk, points, pois = []) {
    const xml = generateKmlXml(walk, points, pois);
    const safeName = (walk.title || "route").replace(/[^a-zA-Z0-9_-]/g, "_");
    downloadFile(`${safeName}_${walk.id || Date.now()}.kml`, xml, "application/vnd.google-earth.kml+xml");
}
